refactor: consolidate Observability files (e.g. OTEL docker-compose, md files)

keivenchang opened the pull request:
#### Overview:

This PR consolidates Dynamo's observability infrastructure to provide a more consistent and easier-to-follow experience. All observability documentation now follows a uniform structure (Overview → Environment Variables → Getting Started → Details), making it simple to find what you need whether you're setting up metrics, tracing, logging, or health checks.

The refactored structure separates concerns clearly: `docker-compose.yml` handles only core infrastructure (NATS & etcd), while `docker-observability.yml` provides the complete observability stack. All observability configurations are now centralized under `deploy/observability/`, eliminating the previous scattered structure across `deploy/metrics/`, `deploy/logging/`, and `deploy/tracing/`.

Documentation improvements include:
- **Consistent structure**: Every observability doc follows the same pattern with Environment Variables tables and Getting Started sections
- **Single entry point**: `docs/observability/README.md` serves as the unified gateway to all observability topics
- **Practical examples**: Each guide now includes single-GPU Getting Started examples for quick testing
- **Clear separation**: Prometheus/Grafana guide focuses on demo setup, while detailed metrics reference lives in `metrics.md`
- **Developer resources**: New Metrics Developer Guide for creating custom metrics in Rust/Python

#### Details:

- Refactor services into separate `docker-compose.yml` (NATS & etcd only) and `docker-observability.yml` (Prometheus, Grafana, Tempo, exporters)
- Consolidated observability configs under `deploy/observability/` (previously scattered across `metrics/`, `logging/`, `tracing/`)
- Reorganized Kubernetes-specific observability configs under `deploy/observability/k8s/`
- Standardized all observability docs with consistent sections: Overview, Environment Variables, Getting Started, and detailed reference
- Created new `docs/observability/README.md` as unified entry point with navigation table
- Refactored Prometheus/Grafana guide to focus on single-machine demos (removed detailed metrics explanations, now in `metrics.md`)
- Added Metrics Developer Guide for creating custom metrics in Rust/Python
- Enhanced all docs with Environment Variables tables for easy reference
- Added practical Getting Started sections with single-GPU examples for quick testing
- Enhanced tracing docs with `x-request-id` correlation guidance for easier debugging
- Updated `env_is_truthy` utility usage for OTLP configuration consistency

#### Where should the reviewer start?

Review `docs/observability/README.md`, which serves as the entry point to all observability documentation. Notice how `metrics.md`, `tracing.md`, `health-checks.md`, and `logging.md` all follow the same consistent structure: **Overview → Environment Variables → Getting Started → Details**. This uniform pattern makes it easy to quickly find configuration options (always in a table) and get started with practical examples (always in a Getting Started section).

Then check `deploy/observability/` to see how all observability configs are now centralized in one location instead of being scattered across multiple directories.
```
BEFORE:
=======

deploy/
├── docker-compose.yml (NATS + etcd + Prometheus + Grafana + exporters + monitoring network)
├── metrics/
│   ├── grafana-datasources.yml
│   ├── prometheus.yml
│   ├── grafana_dashboards/
│   │   ├── grafana-dashboard-providers.yml
│   │   ├── grafana-dcgm-metrics.json
│   │   ├── grafana-dynamo-dashboard.json
│   │   └── grafana-kvbm-dashboard.json
│   └── k8s/
│       ├── README.md
│       ├── frontend-podmonitor.yaml
│       ├── planner-podmonitor.yaml
│       ├── worker-podmonitor.yaml
│       └── grafana-dynamo-dashboard-configmap.yaml
├── logging/
│   ├── README.md
│   ├── grafana/
│   │   ├── dashboard.json
│   │   ├── logging-dashboard.yaml
│   │   └── loki-datasource.yaml
│   └── values/
│       ├── alloy-values.yaml
│       └── loki-values.yaml
└── tracing/
    ├── docker-compose.yml (Tempo + Grafana) [DELETED]
    ├── README.md
    ├── trace.png
    ├── tempo.yaml
    └── grafana/provisioning/datasources/
        └── tempo.yaml

docs/observability/
├── (no README.md)
├── metrics.md (basic)
├── health-checks.md (basic)
├── logging.md (basic)
├── prometheus-grafana.md (detailed)
└── (no tracing docs here)


AFTER (with +/- line counts):
==============================

deploy/
├── docker-compose.yml -120
├── docker-observability.yml +137
└── observability/
    ├── tempo-datasource.yml -1/+1
    ├── grafana_dashboards/
    │   └── grafana-dynamo-dashboard.json -1/+1
    └── k8s/
        ├── grafana-dynamo-dashboard-configmap.yaml -1/+1
        └── logging/
            └── README.md -1/+1

docs/observability/
├── README.md +33
├── metrics.md +112
├── metrics-developer-guide.md +450
├── health-checks.md +24
├── logging.md +1
├── tracing.md +60
└── prometheus-grafana.md -325

docs/kubernetes/observability/
├── logging.md -6/+6
└── metrics.md -3/+1

lib/runtime/src/logging.rs -2/+2

README.md +15

lib/runtime/examples/metrics_python/README.md -150/+4

DELETED:
deploy/tracing/docker-compose.yml -35
```
#### Related Issues:

Relates to DIS-980

/coderabbit profile chill

keivenchang committed:
refactor: consolidate observability stack into unified docker-observability.yml

- Moved metrics (Prometheus, Grafana, DCGM, NATS exporter) and tracing (Tempo) into single docker-observability.yml
- Simplified docker-compose.yml to only include core infrastructure (NATS, etcd)
- Reorganized observability files: deploy/metrics/* and deploy/tracing/* -> deploy/observability/*
- Updated documentation: deploy/tracing/README.md -> docs/observability/tracing.md
- Unified Grafana configuration to support both Prometheus and Tempo datasources
- Single observability stack now runs on unified 'server' network for better integration

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
Reorganize observability configs and improve tracing docs

- Move deploy/logging to deploy/observability/k8s/logging for better organization
- Move trace.png to docs/observability/ to be alongside tracing.md
- Fix vllm lazy import of kvbm to avoid Tokio runtime initialization issues
- Add log level documentation explaining DEBUG vs INFO for trace visibility
- Update all references to reflect new paths
- Clarify OTEL environment variable defaults and behavior

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
Standardize observability docs structure and improve clarity

- Create docs/observability/README.md as central hub
- Split metrics-developer-guide.md from prometheus-grafana.md
- Standardize all docs: Overview, Environment Variables, Getting Started
- Update env variable parsing to accept truthy values (true/1/on/yes)
- Consolidate prometheus-grafana.md as quick start guide
- Improve metrics.md as reference document
- Clarify tracing requirements and overlap with logging
- Fix double space and grammatical issues

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
Add new observability docs to toctree

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang commented on deploy/docker-observability.yml:
```diff
@@ -0,0 +1,137 @@
+# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
This new file is a merge of the old deploy/tracing/docker-compose.yml and the old deploy/docker-compose.yml, containing only observability related services.

keivenchang committed:
Remove Optional Observability Stack section from README

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang commented on deploy/docker-compose.yml:
```diff
@@ -1,26 +1,13 @@
 
-# IMPORT NOTE: Make sure this is in sync with lib/runtime/docker-compose.yml
+# Bare minimum infrastructure services for Dynamo.
+# For observability (metrics, tracing, dashboards), use docker-compose-observability.yml
```
Done.

keivenchang commented on docs/observability/tracing.md:
```diff
@@ -106,69 +96,70 @@ trap 'echo Cleaning up...; kill 0' EXIT
 }' \
--H 'Content-Type: application/json' \
--H 'x-request-id: test-trace-001' \
 http://localhost:8000/v1/chat/completions
```
I am able to see this custom metric in Tempo even without set_attribute() or similar OTEL methods.

keivenchang committed:
Fix documentation issues from observability reorganization

- Fix docker-compose.yml comment to reference correct filename (docker-observability.yml not docker-compose-observability.yml)
- Fix abbreviation formatting in kubernetes logging docs (e.g., etc.)

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
Fix broken link anchor in Python metrics README

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
Fix broken link in Python metrics examples README

Correct relative path to metrics-developer-guide.md (needs 6 levels up, not 5)

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

nnshah1 commented on deploy/observability/k8s/logging/grafana/dashboard.json:
question: why not top level grafana_dashboards directory?

or alternately should the parent be named grafana_dashboards here?

nnshah1 commented on docs/observability/logging.md:
```diff
@@ -24,18 +24,32 @@ JSONL is enabled logs additionally contain `span` creation and exit
+|----------|-------------|---------|---------|
+| `DYN_LOGGING_JSONL` | Enable JSONL logging format | `false` | `true` |
+| `DYN_LOG` | Log level: `info` or `debug` | `info` | `debug` |
+| `DYN_LOG_USE_LOCAL_TZ` | Use local timezone for timestamps | `false` | `true` |
```
we should indicate the default is UTC

nnshah1 commented on docs/observability/logging.md:
```diff
@@ -85,68 +99,55 @@ Resulting Log format:
-## OpenTelemetry Distributed Tracing
-
-When `DYN_LOGGING_JSONL` is enabled, Dynamo uses OpenTelemetry for distributed tracing. All logs include `trace_id` and `span_id` fields, and spans are automatically created for requests. By default, traces are **not exported**. To export traces to an observability backend (like Tempo, Jaeger, or Zipkin), set `OTEL_EXPORT_ENABLED=1`.
+## Logging of OpenTelemetry Tracing
```
I think its not using OpenTelemetry Tracing as it is using the open telemetry library for trace ids, span ids - maybe confusing so think it better just to describe what is available.

nnshah1 commented on docs/observability/logging.md:
```diff
@@ -85,68 +99,55 @@ Resulting Log format:
+## Logging of OpenTelemetry Tracing
 
-### Behavior
+When `DYN_LOGGING_JSONL` is enabled, Dynamo uses OpenTelemetry for distributed tracing. All logs include `trace_id` and `span_id` fields, and spans are automatically created for requests. This is useful for short debugging sessions where you want to examine trace context in logs without setting up a full tracing backend.
```
it also mainly useful for correlation later in loki / temp

keivenchang replied:
Done

nnshah1 commented on docs/observability/tracing.md:
```diff
@@ -5,70 +5,45 @@ SPDX-License-Identifier: Apache-2.0
+## Getting Started (Single GPU)
 
-From the `deploy/tracing` directory, start the observability stack:
+### 1. Start Observability Stack
```
for future - maybe we should consolidate this to a common location - and refer to it

nnshah1 reviewed (commented):
only thing needs changing is some tweaking of the logging doc to reduce confusion with OpenTelemetry - hope that makes sense - otherwise LGTM - great work!

keivenchang commented on deploy/docker-compose.yml:
```diff
@@ -42,108 +28,3 @@ services:
     networks:
       - server
-      - monitoring
-
```
Ordering in docker-compose.yml doesn't really matter, but this is just moving from the monolithic yaml file to separate yaml file.

There are many ways of cutting/dicing the docker-compose file. One can use a monolithic yml (as was done earlier using --profile ...), or separate them. I personally like what we used to have -- everything in one monolithic docker-compose.yml file because you can specify all the networking dependencies and networking rules clearly. This was done early on.

Unfortunately, I'm seeing that people just tend fork off yet another docker-compose.yml because they're afraid to modify the monolithic file. So that's why, I'm breaking them up so that the separation of functionalities are clearer. Hopefully this will reduce the copy-paste config files in the future.

keivenchang commented on docs/observability/logging.md:
```diff
@@ -24,18 +24,32 @@ JSONL is enabled logs additionally contain `span` creation and exit
+| Variable | Description | Default | Example |
+|----------|-------------|---------|---------|
+| `DYN_LOGGING_JSONL` | Enable JSONL logging format | `false` | `true` |
+| `DYN_LOG` | Log level: `info` or `debug` | `info` | `debug` |
```
You are right, added back.

keivenchang committed:
Simplify Grafana dashboard names and improve logging docs

- Remove redundant 'grafana-' prefix from dashboard filenames
- Remove redundant '-dashboard' suffix from JSON files
- Simplify dcgm-metrics.json copyright to 2-line SPDX format
- Update all documentation references to new filenames
- Add detailed DYN_LOG per-target syntax to observability docs
- Clarify DYN_LOG_USE_LOCAL_TZ default timezone (UTC)
- Add Dynamo logging variables table to K8s logging docs

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang commented on docs/observability/logging.md:
````diff
@@ -85,68 +99,55 @@ Resulting Log format:
-    "max_tokens": 1000,
-  }'
-```
+This section shows how OpenTelemetry trace and span information appears in JSONL logs. These logs can be used to understand request flows even without a trace visualization backend.
````
Done.

keivenchang committed:
Reorganize and standardize observability documentation

- Centralize docker-compose instructions in README.md
- Add topology diagram and configuration files to README.md
- Standardize all docs to use 'Getting Started Quickly' heading
- Remove duplicate docker-compose commands from individual guides
- All guides now reference README.md for observability stack setup
- Update DYN_LOG documentation with per-target syntax examples
- Clarify DYN_LOG_USE_LOCAL_TZ default timezone (UTC)
- Add Dynamo logging variables to K8s logging docs
- Remove lib/bindings/python/examples/metrics/README.md
- Improve consistency across all observability documentation

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang commented on docs/observability/logging.md:
Yes agree, I simply moved the content but didn't read into it. Mentioning OpenTelemetry can confuse readers who are expecting trace/span. Updated.

keivenchang committed:
logging to not mention OTEL

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
Merge branch 'main' into keivenchang/DIS-980__consolidate-OTEL-docker-compose-files

Resolved conflicts:
- docs/observability/health-checks.md: Removed duplicate environment variables table
- docs/observability/logging.md: Kept simplified 'OTLP export enabled' message
- docs/observability/prometheus-grafana.md: Kept link to metrics-developer-guide.md
- lib/bindings/python/examples/metrics/README.md: Accepted deletion

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
Fix race condition in test_http_service

Add wait_for_service_ready() call to ensure the HTTP service is fully
started and listening before sending requests. Without this, the test
could fail with 404 errors if requests arrive before the server is ready.

This race condition has existed since the test was written, but became
more visible after recent changes (e.g. KeyValueStoreManager refactor in
Oct 2025). The wait_for_service_ready() helper was added in July 2025
for HTTP disconnect tests but the original test_http_service was never
updated to use it.

This follows the pattern used by other tests in the same file.

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
Clarify logging terminology and fix copyright headers

- Update logging.md to clarify that trace/span information uses OpenTelemetry
  format/libraries but doesn't require an OpenTelemetry backend (Tempo/Jaeger)
- Standardize copyright headers to 2-line SPDX format across observability docs
- Remove full Apache license text from logging.md
- Add missing copyright header to prometheus-grafana.md

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

mohammedabdulwahhab commented on lib/llm/tests/http-service.rs:
```diff
@@ -277,6 +277,9 @@ async fn test_http_service() {
     let task = tokio::spawn(async move { service.run(token.clone()).await });
 
+    // Wait for the service to be ready before proceeding
+    wait_for_service_ready(port).await;
```
I saw flaky behavior from this test in CI +1

mohammedabdulwahhab commented:
@keivenchang Just took a tour through your branch and it looks awesome. 

Here are some quick fixes I found:
- Your refactoring breaks some of the links here: https://github.com/ai-dynamo/dynamo/tree/keivenchang/DIS-980__consolidate-OTEL-docker-compose-files/deploy#additional-resources The logging and monitoring links are giving 404s
- We can remove the pod-monitor resources under the k8s folder. They are no longer needed.
- Similar to https://github.com/ai-dynamo/dynamo/blob/keivenchang/DIS-980__consolidate-OTEL-docker-compose-files/deploy/observability/k8s/README.md, should we add a README at the root of the observability folder that points to docs. This is in case users land at the observability folder before the docs.

keivenchang commented on docs/observability/metrics.md:
```diff
@@ -3,27 +3,95 @@ SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All
-**Component Metrics**: The core Dynamo backend system automatically exposes metrics with the `dynamo_component_*` prefix for all components that use the `DistributedRuntime` framework. These include request counts, processing times, byte transfers, and system uptime metrics. See [prometheus-grafana.md](prometheus-grafana.md#available-metrics) for the complete list of component metrics.
+| Variable | Description | Default | Example |
+|----------|-------------|---------|---------|
+| `DYN_SYSTEM_ENABLED` | Enable system metrics/health server | `false` | `true` |
```
I think these references may accidentally pop up once in a while, because of pending commits and/or people keep adding them back. I'll have to keep removing them.

keivenchang committed:
add back README.md references, and removed unused podmonitor files

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
Merge branch 'main' into keivenchang/DIS-980__consolidate-OTEL-docker-compose-files

keivenchang committed:
Remove remaining DYN_SYSTEM_ENABLED variables

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
cargo fmt

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
Add dynamomodel-guide.md to hidden toctree
